crates/stages/src/pipeline/set.rs

```diff
@@ -0,0 +1,207 @@
+    fn default() -> Self {
+        Self { stages: HashMap::new(), order: Vec::new() }
+    }
+}
+
+impl<DB: Database> Debug for StageSetBuilder<DB> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("StageSetBuilder")
+            .field("stages", &self.stages)
+            .field("order", &self.order)
+            .finish()
+    }
+}
+
+impl<DB> StageSetBuilder<DB>
+where
+    DB: Database,
+{
+    fn index_of(&self, stage_id: StageId) -> usize {
+        let index = self.order.iter().position(|&id| id == stage_id);
+
+        index.unwrap_or_else(|| panic!("Stage does not exist in set: {stage_id}"))
+    }
+
+    fn upsert_stage_state(&mut self, stage: Box<dyn Stage<DB>>, added_at_index: usize) {
+        let stage_id = stage.id();
+        if self.stages.insert(stage.id(), StageEntry { stage, enabled: true }).is_some() {
+            if let Some(to_remove) = self
+                .order
+                .iter()
+                .enumerate()
+                .find(|(i, id)| *i != added_at_index && **id == stage_id)
+                .map(|(i, _)| i)
+            {
+                self.order.remove(to_remove);
+            }
+        }
+    }
+
+    /// Overrides the given [`Stage`], if it is in this set.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the [`Stage`] is not in this set.
+    pub fn set<S: Stage<DB> + 'static>(mut self, stage: S) -> Self {
+        let entry = self
+            .stages
+            .get_mut(&stage.id())
+            .unwrap_or_else(|| panic!("Stage does not exist in set: {}", stage.id()));
+        entry.stage = Box::new(stage);
+        self
+    }
+
+    /// Adds the given [`Stage`] at the end of this set.
+    ///
+    /// If the stage was already in the group, it is removed from its previous place.
+    pub fn add_stage<S: Stage<DB> + 'static>(mut self, stage: S) -> Self {
+        let target_index = self.order.len();
+        self.order.push(stage.id());
+        self.upsert_stage_state(Box::new(stage), target_index);
+        self
+    }
+
+    /// Adds the given [`StageSet`] to the end of this set.
+    ///
+    /// If a stage is in both sets, it is removed from its previous place in this set. Because of
+    /// this, it is advisable to merge sets first and re-order stages after if needed.
+    pub fn add_set<Set: StageSet<DB>>(mut self, set: Set) -> Self {
+        for stage in set.build().finish() {
+            let target_index = self.order.len();
+            self.order.push(stage.id());
+            self.upsert_stage_state(stage, target_index);
+        }
+        self
+    }
+
+    /// Adds the given [`Stage`] before the stage with the given [`StageId`].
+    ///
+    /// If the stage was already in the group, it is removed from its previous place.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the dependency stage is not in this set.
+    pub fn add_before<S: Stage<DB> + 'static>(mut self, stage: S, before: StageId) -> Self {
```

Comment:
are we actually using all these special insert functions?

Reply:
This is with crate consumers in mind. Imagine you want to build on top of reth, adding an indexing stage. Instead of having to figure out what the order of all stages in reth are, you can just use the default stage set and insert your indexing stage after execution.